### ISSUE TITLE
Virtual Keyboard not appear in WPF

### ISSUE THREAD

**Issue by BriceJantieu** [cefsharp/CefSharp#1736]
Hello, 

I've some issues with the virtual keyboard in the WebView on a Dell Optiplex 7440 used in tablet mode, without physical keyboard plugged.
- **What version of the product are you using?**
  - NuGet 39.0.2 / 47.0.2 / 47.0.4 / 49.0.0 / 49.0.1
- **What architecture x86 or x64?**
  - x64
- **On what operating system?**
  - Win10
- **Are you using `WinForms`, `WPF` or `OffScreen`?**
  - WPF
- **What steps will reproduce the problem?**
  - Simply create a webview, go on google and touch the search bar (or any input text of any website)
- **What is the expected output? What do you see instead?**
  - The native virtual keyboard of Win10 should appear. Instead, nothing happen.
- **Please provide any additional information below.**
  - The application must be used with touch screen. No physical keyboard will finally be pluged. So, when I touch a search bar or a textbox in the webview, the native virtual keyboard should appear, like it does when using Chrome or any app where text can be entered. I can only make the virtual keyboard to appear in WinForm example, but not with the WPF one...
- **Does this problem also occur in the `Cef` TestApp from http://cefbuilds.com?**
  - Yes, but only in the webview; in the url bar, the virtual keyboard appear.

**Comment by amaitland** [comment]
The `WPF` control has very limited touch support.

**Comment by bbriot** [comment]
I meet exactly the same problem describe by BriceJantieu:

**What version of the product are you using?**
        NuGet 39.0.2 / 47.0.2 / 47.0.4 / 49.0.0 / 49.0.1 / 51.0

**What architecture x86 or x64?**
        x86

** On what operating system?**
        Win10

**Are you using WinForms, WPF or OffScreen?**
        WPF

Have you got an idea of the roadmap of this issue ?

**Comment by amaitland** [comment]
> Have you got an idea of the roadmap of this issue ?

Best bet is looking into the issue further yourself :+1:

**Comment by johnoso27** [comment]
I  have the same problem: Virtual Keyboard not appear in WPF
What version of the product are you using?
NuGet: 49.0.1 / 51.0
What architecture x86 or x64?
x86

On what operating system?
Win10

Are you using WinForms, WPF or OffScreen?
WPF

Has it solution?

Thanks..

**Comment by amaitland** [comment]
I've never tried to implement a solution, so this is just a rough. Try using the`IRenderProcessMessageHandler.OnFocusedNodeChanged` feature to implement one.

- Enable `FocusedNodeChangedEnabled` https://github.com/cefsharp/CefSharp/blob/cefsharp/51/CefSharp.Example/CefExample.cs#L165
- Implement `IRenderProcessMessageHandler.OnFocusedNodeChanged` https://github.com/cefsharp/CefSharp/blob/cefsharp/51/CefSharp.Example/RenderProcessMessageHandler.cs#L13 Show/hide your keyboard as required based on the node type(check the attributes).

There is no example, so you'd have to write the book on this one. I have no idea how to reliable trigger the keyboard, you'll have to investigate that yourself also.

**Comment by johnoso27** [comment]
Thanks amaitland,

I resolved it implementing IRenderProcessMessageHandler.OnFocusedNodeChanged: manually opening the virtual keyboard when the control focused is INPUT and its type is to capturing something.

 public void OnFocusedNodeChanged(IWebBrowser browserControl, IBrowser browser, IFrame frame, IDomNode node)
        {
            if (this.configuracion != null && this.configuracion.HabilitarTecladoVirtual && this.configuracion.TiposInputPermitidos.Any())
            {
                if (File.Exists(this.configuracion.RutaTeclado))
                {
                    if (node != null &&
                        node["type"] != null &&
                        !string.IsNullOrWhiteSpace(node.TagName) &&
                        node.TagName.ToUpper() == "INPUT" &&                        
                        this.configuracion.TiposInputPermitidos.Any(t => t.ToUpper() == node["type"].ToUpper()))
                    {
                        Process.Start(this.configuracion.RutaTeclado);
                        this.Dispatcher.BeginInvoke((Action)delegate { this.webBrowser.Focus(); });
                    }
                    else
                    {
                        var procesos = Process.GetProcessesByName(this.configuracion.NombreProcesoTeclado);
                        foreach (var proceso in procesos)
                        {
                            try
                            {
                                proceso.Kill();
                            }
                            catch
                            {
                                //// Se omite el error.
                            }
                        }
                    }
                }
                else
                {
                    Aplicacion.Instancia.LogProxy.GuardarNotificacion("La ruta del teclado virtual es inválida.", TipoNotificacion.Error, PrefijoLog);
                }
            }
        }

**Comment by jeffxpx** [comment]
I got in touch with someone at Microsoft about the new virtual keyboard support in WPF 4.6.2 and here's what they said:

> If you have Windows 10 Anniversary Edition (or higher) and are running in tablet mode (WPF respects the settings of the touch KB itself) and touch a control that can get keyboard focus and reports itself as implementing the text UIAutomation pattern, then you should see the keyboard pop up.
> 
> Your more specific question of the type of KB shown has to do with the InputScope property on a control.  You can set InputScope to one of [these values](https://msdn.microsoft.com/en-us/library/system.windows.input.inputscopenamevalue(v=vs.110).aspx) and it should react to your specification.  
> 

I'm going to play around with this over the coming days / weeks and let you know what I find.  Thankfully for testing I can just subclass ChromiumWebBrowser and override OnCreateAutomationPeer and see what sort of stuff I can make work.  Once I get it into a good enough state, I'll take the code I put in the subclass and add it to ChromiumWebBrowser and send over a PR.

**Comment by johnoso27** [comment]
Thanks, I tested DisableWPFTabletSupport method recomended by microsoft to show the virtual keyboard, but as concecuence fails the touch scroll. I can´t update the framework .Net, the app is developed based in framework 4.

**Comment by mismaiel62** [comment]
Hi,
I am using cefsharp version 57 and running chromium in kiosk mode in tablet.
There are some edit control box and want to launch virtual keyboard on selecting any edit box.
is there any way around in chromium for this?
Thanks in advance ;)

**Comment by amaitland** [comment]
For reference a new upstream PR was just opened https://bitbucket.org/chromiumembedded/cef/pull-requests/202/added-cefrenderhandler/diff

**Comment by amaitland** [comment]
New method was added to `CEF` in commit https://bitbucket.org/chromiumembedded/cef/commits/cb793a1d9ba57ec3125e273d0d1406ec90461898

`CefRenderHandler::OnVirtualKeyboardRequested` should provide notification of when a `Virtual Keyboard` should appear and it's type. Unclear at this point how to integrate that with `WPF`.

If anyone has an idea then suggestions welcome.

**Comment by amaitland** [comment]
`OnVirtualKeyboardRequested` was added in #2689

Currently there is no default implementation, you will have to inherit `ChromiumWebBrowser` and override `OnVirtualKeyboardRequested` to implement one for now.

https://github.com/maximcus/WPFTabTip looks promising, will need to investigate further.

**Comment by jeffxpx** [comment]
@amaitland The way WPFTabTip does it (opening and killing tabtip.exe) will lead to lots of sadness.  You want to use the InputPane APIs from the Windows Runtime.

See: https://github.com/Microsoft/WPF-Samples/blob/master/Input%20and%20Commands/TouchKeyboard/TouchKeyboardNotifier/InputPaneRcw.cs

and https://github.com/Microsoft/WPF-Samples/blob/master/Input%20and%20Commands/TouchKeyboard/TouchKeyboardNotifier/TouchKeyboardEventManager.cs#L340..L350 for how to get an instance of an InputPane

Once you have an instance, you can use TryShow() and TryHide().

https://docs.microsoft.com/en-us/uwp/api/windows.ui.viewmanagement.inputpane.tryshow

**Comment by amaitland** [comment]
> You want to use the InputPane APIs from the Windows Runtime.

@jeffxpx How does this behave on `Windows 7`?

> Once you have an instance, you can use TryShow() and TryHide().
> 
> https://docs.microsoft.com/en-us/uwp/api/windows.ui.viewmanagement.inputpane.tryshow

This looks like `APIs` from `UWP`, not sure we can use those in a general sense as I'm not sure they're supported on all platforms we support, `Win 7, 8.1, 10`.

With version `73` anyone will be able to choose their own implementation.

**Comment by amaitland** [comment]
As `WPF` doesn't have an actual `API` for showing the virtual keyboard, I'm leaving this it up to the user to select the appropriate solution for their needs. If `Microsoft` ever comes to the party then we can include a default implementation. A `VirtualKeyboardRequested` event has been added to make it trivially easy to be notified.

An example of using it will look like 

```c#
private void BrowserVirtualKeyboardRequested(object sender, VirtualKeyboardRequestedEventArgs e)
{
	var inputPane = touchKeyboardEventManager.GetInputPane();

	if (e.TextInputMode == TextInputMode.None)
	{
		inputPane.TryHide();
	}
	else
	{
		inputPane.TryShow();
	}
}
```

I've provided a very basic `Windows 10` example below, others using `Windows 8` may need to use something like https://github.com/maximcus/WPFTabTip

Someone can contribute a `Windows 7` example if they like, I don't have a `Windows 7` machine at the moment, so not really possible.

---

- `VirtualKeyboardRequested` event added in commit https://github.com/cefsharp/CefSharp/commit/cddece0cdca990a6b81c0764c34088ff2a5c6320

- `Windows 10 Only` example added in commit https://github.com/cefsharp/CefSharp/commit/0b57e526158e57e522d46671404c557256529416

Code adapted from https://github.com/Microsoft/WPF-Samples/blob/master/Input%20and%20Commands/TouchKeyboard/TouchKeyboardNotifier/Readme.md (which is thankfully `MIT` licensed).
